consensus/core/src/dag_state.rs

```diff
@@ -307,6 +319,99 @@ impl DagState {
             .set(highest_accepted_round_for_author as i64);
     }
 
+    // Updates votes for certification of transactions in the causal history of the block.
+    // TODO(fastpath): add randomized tests.
+    fn update_certification_votes(&mut self, block: &VerifiedBlock) -> Vec<BlockOutput> {
+        let mut certified_blocks = vec![];
+
+        // When a block has an explicit vote, record the rejections. The rest of the transactions are implicitly accepted.
+        for block_votes in block.transaction_votes() {
+            let Some(block_info) = self.recent_blocks.get_mut(&block_votes.block_ref) else {
+                // TODO(fastpath): ensure the voted block exists in the DAG, with BlockManager.
+                // If the block is not found, it is outside of GC bound.
+                continue;
+            };
+            block_info
+                .accept_votes
+                .add_unique(block.author(), &self.context.committee);
+            for reject in &block_votes.rejects {
+                block_info
+                    .reject_votes
+                    .entry(*reject)
+                    .or_default()
+                    .add_unique(block.author(), &self.context.committee);
```

Comment:
I think doing this check in BlockManager fits the existing strategy, to avoid processing detectable equivocating blocks as much as possible. Added a TODO.